### PR TITLE
Prefix Cache Aware Scheduling [1/n] 

### PR DESCRIPTION
**Pull request by rickyyx**
# TL;DR
- prototype in https://github.com/vllm-project/vllm/pull/9862
- addresses https://github.com/vllm-project/vllm/issues/7883
- This PR makes scheduling prefix aware in V0. KV Cache utilization will remain high with high prefix cache hit.
- Yielding up to 50% throughput improvement on high prefix cache hit for batch inferencing, and 25% request throughput when online serving.
- It doesn't introduce e2e overheads when the workload doesn share prefix. 

## Background

With current impl in main, there are at least 2 places where scheduling is not optimal:
1. When deciding how many new tokens should be used for scheduling budget, it doesn't take into account already computed tokens. 
2. When deciding if a sequence could be allocated for prefill, it doesn't take into account the already computed blocks.

This would result in under-utilization of KV cache, and un optimal scheduling decision for a batch.

For more details, see the https://github.com/vllm-project/vllm/issues/7883

## High Level Approach
This PR addresses the issue by solving 1 only: 
1.  **Make scheduling prefix cache aware**: When deciding how many tokens to schedule, the already cached tokens are now also taken into account, and the scheduling budget now includes cached tokens. 
2. The `ComputedBlocksTracker` will track the block hashes for a sequence and tell scheduler/block manager how many tokens have been cached given a sequence. 

On a high level, there are below major changes.
1. Scheduler when deciding how many tokens to batch, it also takes into account new tokens that are already cached. So the scheduling budget only includes **new and uncached** tokens. 
2. The scheduler gets the cached token info by asking the blockmanager which has the computed blocks tracker tracks the information for each sequence. 
3. For each sequence, a block hash is computed twice in the current PR (once at the computed block tracker, once in the allocator). A future PR could improve this. 

## Throughput Benchmark
Example command
```
    python benchmarks/benchmark_prefix_caching.py \
        --model meta-llama/Meta-Llama-3-8B-Instruct \
        --enable-prefix-caching \
        --use-v2-block-manager \
        --num-prompts 300 \
        --input-length-range 1000:1000 \
        --output-len 10 \
        --prefix-len $prefix_len \
        --max-num-batched-tokens 2048 \
        --enable-chunked-prefill
```

image

More details in this [doc](https://docs.google.com/spreadsheets/d/1g2IX991AkAa9rdCW5hl2hOf41fZZHlP30SG3-y3kGcE/edit?gid=0#gid=0)

## Serving Benchmark

```
# Server
vllm serve meta-llama/Meta-Llama-3-8B-Instruct --disable-log-requests --enable-prefix-caching --enable-chunked-prefill
```

#### Serving Results on QPS=10 
- The improvement on TTFT is significant. 
- There's slight improvement on request throughput
- There's slight degrade on ITL (likely due to more tokens from prefill in a batch)

```
python3 benchmarks/benchmark_serving.py --backend vllm --model meta-llama/Meta-Llama-3-8B-Instruct --dataset-name random \
--random-input-len 200 --random-output-len 30 --random-prefix-len 600 --seed 0 \
--request-rate 10 --num-prompts 500
```

```
============ Serving Benchmark Result ============
                                          PR           Main
Successful requests:                     500           500       
Benchmark duration (s):                  52.01         55.42     
Total input tokens:                      400000        400000    
Total generated tokens:                  14831         14809     
Request throughput (req/s):              9.61          9.02      
Output token throughput (tok/s):         285.16        267.23    
Total Token throughput (tok/s):          7976.00       7485.27   
---------------Time to First Token----------------
Mean TTFT (ms):                          238.19        2193.42   
Median TTFT (ms):                        209.70        2510.56   
P99 TTFT (ms):                           632.90        3922.97   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          81.33         60.74     
Median TPOT (ms):                        75.03         61.54     
P99 TPOT (ms):                           135.80        64.79     
---------------Inter-token Latency----------------
Mean ITL (ms):                           82.12         61.21     
Median ITL (ms):                         62.17         62.90     
P99 ITL (ms):                           285.48        116.98    
==================================================
```

#### Serving Results on QPS=15
With a higher request rate (QPS=15), the improvement on TTFT and request rate will be more significant (25%)
```
python3 benchmarks/benchmark_serving.py --backend vllm --model meta-llama/Meta-Llama-3-8B-Instruct --dataset-name random --random-input-len 200 --random-output-len 50 --random-prefix-len 600 --seed 0 \
--request-rate 15 \
--num-prompts 500
```

```
============ Serving Benchmark Result ============
                                          PR           Main
Successful requests:                     500           500       
Benchmark duration (s):                  46.94         58.66     
Total input tokens:                      400000        400000    
Total generated tokens:                  24685         24668     
Request throughput (req/s):              10.65         8.52      
Output token throughput (tok/s):         525.92        420.52    
Total Token throughput (tok/s):          9048.08       7239.31   
---------------Time to First Token----------------
Mean TTFT (ms):                          4797.81       11201.85  
Median TTFT (ms):                        4566.95       11076.55  
P99 TTFT (ms):                           10789.80      22585.37  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          160.69        64.68     
Median TPOT (ms):                        170.73        65.03     
P99 TPOT (ms):                           191.34        68.68     
---------------Inter-token Latency----------------
Mean ITL (ms):                           161.50        65.20     
Median ITL (ms):                         154.61        66.75     
P99 ITL (ms):                           847.72        124.48    
==================================================
```

#### Serving Results on No Prefix Shared
```
python3 benchmarks/benchmark_serving.py --backend vllm --model meta-llama/Meta-Llama-3-8B-Instruct --dataset-name random --random-input-len 800 --random-output-len 50 --random-prefix-len 0 --seed 0 --request-rate 10 --num-prompts 500
```

- There's no overhead when workload has no shared prefil.
```
============ Serving Benchmark Result ============
                                          PR           Main         
Successful requests:                     500           500          
Benchmark duration (s):                  127.43        127.65       
Total input tokens:                      400000        400000       
Total generated tokens:                  24350         24350        
Request throughput (req/s):              3.92          3.92         
Output token throughput (tok/s):         191.08        190.76       
Total Token throughput (tok/s):          3330.05       3324.43      
---------------Time to First Token----------------
Mean TTFT (ms):                          36577.91      36610.46     
Median TTFT (ms):                        36444.46      36417.98     
P99 TTFT (ms):                           73755.41      73895.10     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          144.37        144.74       
Median TPOT (ms):                        143.66        143.89       
P99 TPOT (ms):                           302.23        302.97       
---------------Inter-token Latency----------------
Mean ITL (ms):                           142.62        142.98       
Median ITL (ms):                         143.56        143.77       
P99 ITL (ms):                           286.82        287.27       
==================================================
```

### CODE REVIEW

File: vllm/core/scheduler.py [hunk 1]
```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+        else:
+            num_cached_new_tokens_seq = 0
+
+        all_num_new_tokens_seq = seq.get_len() - num_computed_tokens_seq
+        num_uncached_new_tokens_seq = (all_num_new_tokens_seq -
+                                       num_cached_new_tokens_seq)
+
+        return num_uncached_new_tokens_seq, num_cached_new_tokens_seq
+
+    def _chunk_new_tokens_to_schedule(
```
Comment (rickyyx) [inline vllm/core/scheduler.py:1750]: mainly refactor from the main's logic.

File: vllm/core/block/prefix_caching_block.py [hunk 2]
```diff
@@ -634,6 +616,49 @@ def swap_in(self, blocks: List[Block]) -> None:
...
 
+    def find_cached_blocks_prefix(self, block_hashes: List[int]) -> List[int]:
+        """
+        Given a list of block hashes, return the prefix of the block hashes that
+        are all cached.
+
+        Since a block's block hash includes the hashes of all previous blocks,
+        and we only allocate/deallocate blocks in the entire sequence, so if a
+        block is cached, then all previous blocks are also cached. With this
+        property, we can use binary search to find the prefix of cached blocks.
```
Comment (comaniac) [inline vllm/core/block/prefix_caching_block.py:627]: 👍

File: vllm/core/block/prefix_caching_block.py [hunk 3]
```diff
@@ -843,86 +868,124 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
...
+        self._allocator = allocator
+        self._block_size = block_size
+        self._enable_caching = enable_caching
+
+        # A map from seq_id to the list of block hashes for the
+        # sequence. This is so that we don't have to recompute the block hashes
+        # for the sequence when we need to check if the sequence is cached.
+        # Note a block that's not full will not have its hash calculated and
+        # recorded.
+        self._full_blocks_hashes: Dict[int, List[int]] = {}
```
Comment (comaniac) [inline vllm/core/block/prefix_caching_block.py:900]: Might be better naming.
```suggestion
        self._seq_id_to_blocks_hashes: Dict[int, List[int]] = {}
```

File: vllm/core/block/prefix_caching_block.py [hunk 4]
```diff
@@ -843,86 +868,124 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
...
+
+        # A map from seq_id to the number of tokens that are cached for the
+        # sequence.
+        # We need this so that a sequence in continuous prefill doesn't
+        # accidentally see its cached token count change. See comments in
+        # `get_num_cached_tokens` for more details.
+        self._num_tokens_computed: Dict[int, int] = {}
```
Comment (comaniac) [inline vllm/core/block/prefix_caching_block.py:907]: Ditto
```suggestion
        self._seq_id_to_num_tokens_computed: Dict[int, int] = {}
```

File: vllm/core/block/prefix_caching_block.py [hunk 5]
```diff
@@ -843,86 +868,124 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
...
+        cur_num_blocks_recorded = len(block_hashes_recorded)
+        token_ids = seq.get_token_ids()
+        assert len(token_ids) >= cur_num_blocks_recorded * self._block_size, (
+            f"The sequence has {len(token_ids)} tokens, but"
+            f" already recorded {cur_num_blocks_recorded} blocks. "
+            "This should not happen since we assume blocks are "
+            "only appended other than recomputation. When the sequence is "
+            "recomputed, we should have removed the info of the old blocks.")
+        # Update the computed block hashes for the sequence
+        num_total_blocks = len(token_ids) // self._block_size
```
Comment (comaniac) [inline vllm/core/block/prefix_caching_block.py:923]:
```suggestion
                # Update the computed block hashes for the sequence. Since only full blocks are considered as "computed", we take floor here.
        num_computed_blocks = len(token_ids) // self._block_size
```

File: vllm/core/block/prefix_caching_block.py [hunk 6]
```diff
@@ -843,86 +868,124 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
...
+            block_hashes_recorded.append(block_hash)
+            prev_block_hash = block_hash
+
+        self._full_blocks_hashes[seq.seq_id] = block_hashes_recorded
+
+    def get_num_cached_tokens(self, seq: Sequence) -> int:
+        if not self._enable_caching:
+            return 0
+
+        # We always tries to update the sequence hashes for the sequence.
```
Comment (comaniac) [inline vllm/core/block/prefix_caching_block.py:951]:
```suggestion
        # We always try to update the sequence hashes for the sequence.
```

File: vllm/core/scheduler.py [hunk 7]
```diff
@@ -56,24 +56,35 @@ class SchedulingBudget:
...
+    _num_cached_tokens: int = 0
+    # Number of actual non-cached tokens in the batch.
     _num_batched_tokens: int = 0
     _num_curr_seqs: int = 0
 
     def can_schedule(self, *, num_new_tokens: int, num_new_seqs: int):
-        assert num_new_tokens != 0
+        # We allow num_new_tokens to be 0 when the entire sequence has
+        # been cached.
+        assert num_new_tokens >= 0
```
Comment (comaniac) [inline vllm/core/scheduler.py:68]: Actually I se all use cases of this function have a pre-condition `if num_nre_tokens==0 or can_schedule(num_new_tokens)`, so I guess this assertion still works?

File: vllm/core/scheduler.py [hunk 8]
```diff
@@ -541,9 +556,17 @@ def _schedule_running(
         assert len(self._async_stopped) == 0
         while running_queue:
             seq_group = running_queue[0]
-            num_running_tokens = self._get_num_new_tokens(
+            # We discard the cached tokens info here because we don't need it
+            # for running sequence because:
```
Comment (comaniac) [inline vllm/core/scheduler.py:560]:
```suggestion
            # We discard the cached tokens info here because we don't need it
            # for running sequence:
```

File: vllm/core/scheduler.py [hunk 9]
```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
 
         return self.scheduler_config.num_lookahead_slots
 
-    def _get_num_new_tokens(self, seq_group: SequenceGroup,
-                            status: SequenceStatus, enable_chunking: bool,
-                            budget: SchedulingBudget) -> int:
-        """Get the next new tokens to compute for a given sequence group
-            that's in a given `status`.
+    def _get_num_new_tokens_to_schedule(
```
Comment (comaniac) [inline vllm/core/scheduler.py:1639]: It's a bit hard to tell the difference in terms of functionalities between this function and `_get_num_new_tokens_uncached_and_cached` from their names. For example, the term `to_schedule` doesn't convey its functionality but the purpose. Could we try to figure out a better naming for both of them?

File: vllm/core/scheduler.py [hunk 10]
```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+        sequence.
+
+        Returns:
+            A tuple of two ints. The first int is the number of uncached tokens.
+            The second int is the number of cached tokens.
+        """
+        if not seq.is_prefill():
+            # Decode sequences should always just have 1 uncached token
+            # TODO(rickyx): Actually is this still correct for multi-step?
+            return 1, 0
```
Comment (comaniac) [inline vllm/core/scheduler.py:1710]: And speculative decoding?

File: vllm/core/scheduler.py [hunk 11]
```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
...
+                                 f"({budget.token_budget}) % block_size "
+                                 f"({block_size}) = {remainder}")
+            # Round down to block
+            remaining_token_budget = (remaining_token_budget // block_size *
+                                      block_size)
+            # Calculate the number of new tokens that are not cached with
+            # chunk cap.
+            num_new_tokens = min(num_new_tokens, remaining_token_budget)
+        else:
+            num_new_tokens = min(num_new_tokens, remaining_token_budget)
```
Comment (comaniac) [inline vllm/core/scheduler.py:1807]: nit: Avoid repeating the last line, but I don't have strong preference.

```suggestion
        else:
            if self.cache_config.enable_prefix_caching:
                # When prefix caching is enabled, we always allocate
                # the number of new tokens that is dividable by the block
                # size to avoid partial block matching.
                block_size = self.cache_config.block_size
                remainder = budget.token_budget % block_size
                if remainder != 0:
                    raise ValueError("When enabling chunked prefill and "
                                     "prefix caching, max_num_batched_tokens "
                                     "(chunk size) must be dividable by "
                                     "block size, but got chunk_size "
                                     f"({budget.token_budget}) % block_size "
                                     f"({block_size}) = {remainder}")
                # Round down to block
                remaining_token_budget = (remaining_token_budget // block_size *
                                          block_size)
                # Calculate the number of new tokens that are not cached with
                # chunk cap.
            num_new_tokens = min(num_new_tokens, remaining_token_budget)
```

**Review by comaniac — COMMENTED** [review]
Great work! Leave some commits.
Also cc @zhuohan123 @alexm-neuralmagic

File: vllm/core/scheduler.py (hunk 9 above)
Comment (rickyyx) [reply vllm/core/scheduler.py:1639]: Agree - 

Maybe:
- (this) `_get_num_new_tokens_to_schedule` -> `_get_num_new_uncached_and_cached_tokens`
- (the routine below) -> `_get_num_new_uncached_and_cached_tokens_for_seq` ?

File: vllm/core/scheduler.py (hunk 10 above)
Comment (rickyyx) [reply vllm/core/scheduler.py:1710]: hmm I might be missing something - but I think both speculative decoding and multi-step are using some `num_steps` field that's opaque to the scheduler as of now. 

So in a way, doesn't seem the shceduler is multi-step aware.

File: vllm/core/scheduler.py (hunk 9 above)
Comment (comaniac) [reply vllm/core/scheduler.py:1639]: So the current routine below is just used by this one? In this case can we just rename `_get_num_new_tokens_to_schedule` -> `_get_num_new_uncached_and_cached_tokens` and move the routine below to be the inner function?

File: vllm/core/scheduler.py (hunk 7 above)
Comment (rickyyx) [reply vllm/core/scheduler.py:68]: We actually have codes like `if all_new_tokens ==0 or not can_schedule(uncached_new_tokens` with the change now - so sometimes all new tokens count is not 0, but the uncached token count is 0.

Comment (comaniac) [reply vllm/core/scheduler.py:68]: Ah I see. Then we could just improve the comment a bit

```
        # num_new_tokens may be 0 when the entire sequence has
        # been cached.
```

Comment (comaniac) [reply vllm/core/scheduler.py:68]: btw what happen when an entire sequence is cached? Seems like we can always schedule it? But AFAIK our model runner doesn't support query_len=0 so it would be problematic if you send a sequence without input tokens. I'm wondering how this PR overcome this issue?

Comment (rickyyx) [reply vllm/core/scheduler.py:68]: Oh yeah, in this case, the number of cached tokens would be > 0, so the `token_chunk_size` would be > 0. It's just that this sequence doesn't contribute to budgeting anymore, but the new token count should always be >0. I can add an assert somewhere for this. 

But I do see in the main where we never return all blocks as cached when calculating computed blocks before, but I so far haven't seen any issue in my PR. Maybe I can validate it a bit more with some tests.

**Comment by rickyyx** [comment]
Updates
- Added integration test for prefix caching with fully (and partially) cached prompts.

**Comment by rickyyx** [comment]
lol - fml, i hate DCO.

**Review by comaniac — APPROVED** [review]
LGTM. The remaining issue per offline discussion is to make sure each scheduled sequence has at least one token in the budget.

cc @zhuohan123 for another review.

**Comment by rickyyx** [comment]
Updates
- Added test to make sure each scheduled sequence contributes at least one token in the budget (a fully cached prompt should not be scheduled with 0 uncached token).
